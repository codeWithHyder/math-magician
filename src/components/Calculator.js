import React, { useState } from 'react';
import ChildCalculator from './ChildCalculator';
import calculate from '../logic/calculate';

function Calculator() {
  const [displayValue, setDisplayValue] = useState('0');
  const [calculatorData, setCalculatorData] = useState({
    total: null,
    next: null,
    operation: null,
  });

  const handleClick = (buttonName) => {
    const newData = calculate(calculatorData, buttonName);
    setCalculatorData(newData);

    // Update the display value
    const newDisplayValue = newData.next || newData.total || '0';
    setDisplayValue(newDisplayValue);
  };

  const renderButton = (label) => (
    <ChildCalculator
      key={label}
      label={label}
      onClick={() => handleClick(label)}
    />
  );

  return (
    <div className="app-container">
      <input type="text" id="calsc" className="calsc" value={displayValue} readOnly />

      <div className="calculator-panel">
        {['AC', '+/-', '%', 'รท'].map((label) => renderButton(label))}
      </div>
      <div className="calculator-panel">
        {['7', '8', '9', 'x'].map((label) => renderButton(label))}
      </div>
      <div className="calculator-panel">
        {['4', '5', '6', '-'].map((label) => renderButton(label))}
      </div>
      <div className="calculator-panel">
        {['1', '2', '3', '+'].map((label) => renderButton(label))}
      </div>
      <div className="calculator-panel">
        {['0', '.', '='].map((label) => renderButton(label))}
      </div>

    </div>
  );
}

export default Calculator;
