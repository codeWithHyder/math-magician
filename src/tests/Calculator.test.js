import { render, screen, queryByAttribute } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator';

const getById = queryByAttribute.bind(null, 'id');

it('test Calculator component rendering', () => {
  const tree = renderer.create(<Calculator />).toJSON();
  const dom = render(<Calculator />);
  const buttons = screen.getAllByRole('button');
  const value = getById(dom.container, 'bottom-value');
  expect(buttons).toHaveLength(19);
  expect(value?.textContent).toBe('0'); // use optional chaining operator to check for null
  expect(tree).toMatchSnapshot();
});
