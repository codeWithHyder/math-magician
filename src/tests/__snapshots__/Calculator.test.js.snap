// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Calculator component rendering 1`] = `
<div
  className="cal-main-container"
>
  <h2>
    Let us do some Math!
  </h2>
  <div
    className="app-container"
  >
    <input
      className="calsc"
      id="calsc"
      readOnly={true}
      type="text"
      value="0"
    />
    <div
      className="calculator-panel"
    >
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </div>
    <div
      className="calculator-panel"
    >
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </div>
    <div
      className="calculator-panel"
    >
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </div>
    <div
      className="calculator-panel"
    >
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </div>
    <div
      className="calculator-panel"
    >
      <button
        className="calculator-button wide"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        .
      </button>
      <button
        className="calculator-button "
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
